sympy/matrices/matrices.py

```diff
@@ -1141,15 +1153,29 @@ def eigenvals(self, error_when_incomplete=True, **flags):
                     eigs[diagonal_entry] += 1
         else:
             flags.pop('simplify', None)  # pop unsupported flag
-            eigs = roots(mat.charpoly(x=Dummy('x')), **flags)
+            if isinstance(simplify, FunctionType):
+                eigs = roots(mat.charpoly(x=Dummy('x'), simplify=simplify), **flags)
+            else:
+                eigs = roots(mat.charpoly(x=Dummy('x')), **flags)
 
         # make sure the algebraic multiplicty sums to the
         # size of the matrix
         if error_when_incomplete and (sum(eigs.values()) if
             isinstance(eigs, dict) else len(eigs)) != self.cols:
             raise MatrixError("Could not compute eigenvalues for {}".format(self))
 
-        return eigs
+        # Since 'simplify' flag is unsupported in roots()
+        # simplify() function will be applied once at the end of the routine.
+        if not simplify:
+            return eigs
+        if not isinstance(simplify, FunctionType):
+            simplify = _simplify
+        # With 'multiple' flag set true, simplify() will be mapped for the list
+        # Otherwise, simplify() will be mapped for the keys of the dictionary
+        if not multiple:
+            return dict(map(lambda item: (simplify(item[0]), item[1]), eigs.items()))
```

Comment:
Could this be something like: `{simplify(key): value for key, value in eigs.items()}?

Reply:
@jksuom 
Would there be any specific reason to use `for` loop 
than using `map` iterator?

Reply:
I think that this is more readable. Also, the "for loop" is actually a dict comprehension which should be fairly efficient.

Reply:
@jksuom 
If that is the case, I think another line below should also be changed into list comprehension.
How should I proceed?

Reply:
I personally prefer comprehensions but I know that `map` is used in many places in SymPy. In this case, I would accept either version.

Reply:
@jksuom 
Okay, but I will just make it consistent with comprehensions at the moment.

I had originally had thought of coming up with an idea of using `map`, and how it can easily be implemented with multiprocessing.
But in this particular case, I am also seeing `roots` failing with some large scale matrices with random numerics, unable to provide benchmarks at the moment.